### DISCUSSION TITLE
Aggregates: Persist and return stored object

### DISCUSSION THREAD

**Discussion by ahoiroman** [spatie/laravel-event-sourcing#357] [Q&A]
Hello everybody,

I'm quite new to the concept of event sourcing and at the moment I'm playing around in a pretty large app to move to event sourcing on smaller features.

Now I was on my way to implement the aggregates on a small controller like this:

```php
public function store(Request $request): ItemData
    {
        $newUuid = Str::uuid()->toString();

        ItemAggregateRoot::retrieve($newUuid)
                                       ->createItem($request->input('item.name'))
                                       ->persist();

        $item = Item::uuid($newUuid);         // This feels a bit fishy

        return ItemData::from($item);
    }
```

As you can see, I need to return a DTO called `ItemData`, which contains the item I just saved. But the way I am doing this, it kinda looks "fishy" as I am searching for the `Item` I just saved using the UUID.

**Comment by etahamer** [comment]
@ahoiroman As far as I understood, your Item is projection, then I would suggest generating ItemData from ItemAggregateRoot rather than from projection, which may or may not exist during runtime (Happens when projections are being queued).

**Reply by ahoiroman** [reply]
Hi and thank you for your reply.

Unfortunately my frontend does indeed need a response containing data coming from the saved model, such as snowflake id and relations. Maybe a broadcasting of those data would be a better idea…?

**Reply by etahamer** [reply]
Hi, 

In that case I would either broadcast or decouple further functionality from relying on ids. Instead I would use UUID where possible.

**Reply by chrisloftus** [reply]
@ahoiroman I'm having the same problem. Did you go with your original method in the end?

**Reply by ahoiroman** [reply]
I ended up returning the UUID and broadcast the projection‘s data after it has been created.

**Reply by chrisloftus** [reply]
@ahoiroman I'm broadcasting at the moment but it is slow.

**Reply by ahoiroman** [reply]
It may be - but you could make this process visible to your users by reflecting the loading-state in the frontend in a nice way.

**Reply by chrisloftus** [reply]
Maybe but it's still not ideal.

I'm creating a relationship model, not the parent model that has UUID already. So the UUID is on the parent model. So I need to think of a way to return the child model that is created. Maybe use UUIDs on the child model too but ugh...

**Reply by ahoiroman** [reply]
So you asked me for my solution - I gave you the answer that worked for me.

you are not forced to queue the events - you can always return the full model as shown in my initial post.
